### Setting up for Analysis

In [1]:
# Importing packages we could need
import pandas as pd
import os
import numpy as np

In [2]:
# Gets current working directory
os.getcwd()

'c:\\Users\\bfran\\Ironhack\\Week 3\\Project\\Ironhack-PRY-BRZ-MGA'

### Explanation of Data

This dataset we are using compares Paraguay and Brazil reports of Brazil-directed trade.

In layman's terms, it explores the differences between what Paraguay reported exporting to Brazil and what Brazil reported as having imported from Paraguay.

Each row in this dataset specifically examines trade reporting for a single HS code. HS codes are basically a classification of items that most countries generally agree to adhere to. We examined codes at the most specific level, HS 6. The HS 6 level basically distinguishes differences between, say, cardstock colored paper and ruled loose-leaf paper. This level of specificity in examination allows for the greatest level of focus on item based trade fraud. 

Comparing this reported data side-by-side allows a researcher to essentially play a game of 'spot the differences' between the two reports. 

In [3]:
# Saves as df previously created mx_compare.csv file 
df = pd.read_csv('mx_compare.csv')

In [4]:
df

hs_code     pry_x_q pry_x_unit     brz_m_q brz_m_unit  pry_x_net_wgt  \
0      20130  21523200.0         kg  20859060.0         kg     21523200.0   
1      20220     52442.3         kg     52443.0         kg        52442.3   
2      20230   7425400.0         kg   7144609.0         kg      7425400.0   
3      20622   2883500.0         kg   2692900.0         kg      2883500.0   
4      20629   1037276.0         kg    978823.0         kg      1037276.0   
..       ...         ...        ...         ...        ...            ...   
791   851230         0.0         NR         2.0          u            0.0   
792   870810         0.0         NR        15.0         kg            0.0   
793   870894         0.0         NR         2.0         kg            0.0   
794   902610         0.0         NR       102.0          u            0.0   
795   902910         0.0         NR         2.0          u            0.0   

     brz_m_net_wgt   pry_x_value  itic_rate  applied_tariff   adj_x_value  \
0      20859060.00  1.249529e+08       1.01            9.60  1.265773e+08   
1         52443.00  1.404194e+05       1.01            8.00  1.417300e+05   
2       7144609.00  3.770901e+07       1.01            9.60  3.806096e+07   
3       2692900.00  3.287412e+06       1.03            8.00  3.397540e+06   
4        978823.00  2.882343e+06       1.03            8.00  2.978902e+06   
..             ...           ...        ...             ...           ...   
791           0.19  0.000000e+00       1.03           18.00  0.000000e+00   
792          15.00  0.000000e+00       1.01           18.00  0.000000e+00   
793           2.00  0.000000e+00       1.01           15.09  0.000000e+00   
794           8.48  0.000000e+00       1.01           11.80  0.000000e+00   
795           7.20  0.000000e+00       1.02           14.60  0.000000e+00   

     brz_m_value   trade_gap  est_tax_loss    wgt_gap value_ratio wgt_ratio  \
0    120517874.0  6059424.71     581704.77  664140.00        0.95      0.97   
1       140643.0     1087.01         86.96      -0.70        0.99       1.0   
2     36218864.0  1842097.60     176841.37  280791.00        0.95      0.96   
3      3053319.0   344221.49      27537.72  190600.00         0.9      0.93   
4      2749729.0   229172.92      18333.83   58453.00        0.92      0.94   
..           ...         ...           ...        ...         ...       ...   
791          4.0       -4.00         -0.72      -0.19         NXR       NXR   
792         10.0      -10.00         -1.80     -15.00         NXR       NXR   
793         25.0      -25.00         -3.77      -2.00         NXR       NXR   
794       1278.0    -1278.00       -150.80      -8.48         NXR       NXR   
795       1063.0    -1063.00       -155.20      -7.20         NXR       NXR   

     density_ratio  
0             0.98  
1             0.99  
2             0.99  
3             0.96  
4             0.98  
..             ...  
791            NaN  
792            NaN  
793            NaN  
794            NaN  
795            NaN  

[796 rows x 18 columns]

### Creating Subsets

One thing we noticed in this dataset was that there were a high number of rows where one party reported trade activity but the other did not. 

For instance, in line 792, Paraguay reported that it did not send any weight or quantity of HS item 870810, but Brazil reported that they received 15kg of that same item. 

That lack of reporting by one party undermines our ability to analyze the accuracy of either country's reporting. Imagine trying to play a game of 'spot the difference' when you only have one picture, not two.

In order to better analyze and explore our data, we need to separate the rows where there was reporting from both countries from those where there was not. 

In [5]:
# Creates two subsets

# This subset removes rows in which there were no reported import or export weights or values
df_full_rep = df[(df['value_ratio'] != "NXR") 
                 & (df['value_ratio'] != "NMR") 
                 & (df['wgt_ratio'] != "NXR") 
                 & (df['wgt_ratio'] != "NMR")]


# This subset includes only those rows with incomplete reporting.
df_missing_rep = df[~((df['value_ratio'] != "NXR") 
                      & (df['value_ratio'] != "NMR") 
                      & (df['wgt_ratio'] != "NXR") 
                      & (df['wgt_ratio'] != "NMR"))]

In [6]:
pd.set_option('display.max_rows', None)
df_full_rep['wgt_ratio'].value_counts()

wgt_ratio
1.0         180
0.99         27
0.97         16
0.98         13
1.01         11
0.96          7
0.75          5
0.94          5
0.93          5
1.02          5
0.81          4
1.04          4
0.04          3
1.05          3
0.95          3
1.1           3
1.03          3
0.87          3
0.92          3
0.91          3
0.67          2
1.11          2
0.02          2
0.6           2
0.55          2
0.82          2
0.1           2
0.9           2
0.89          2
1.29          2
0.54          2
0.06          2
1.06          2
0.77          2
0.7           2
1.13          2
0.26          2
1.15          2
3.88          1
1.48          1
0.17          1
11.0          1
2.21          1
2.02          1
0.37          1
0.3           1
1.4           1
1.25          1
0.39          1
0.01          1
6.67          1
1.93          1
0.71          1
0.52          1
0.78          1
0.68          1
1.27          1
3.11          1
3.37          1
214.24        1
6.27          1
0.22          

In [7]:
# Prints the shape of each dataset
print("Shape of full reporting data set: ",df_full_rep.shape)
print("Shape of missing reporting data set: ",df_missing_rep.shape)

Shape of full reporting data set:  (414, 18)
Shape of missing reporting data set:  (382, 18)


This demonstrates that trade reporting was definitely a major problem when examining Brazil-directed trade with Paraguay. 

In [8]:
# Counts the occurrences of unique values in value ratio column
df_missing_rep['value_ratio'].value_counts()

value_ratio
NMR     356
NXR      23
0.01      2
0.04      1
Name: count, dtype: int64

However, contrary to our stated hypothesis, the value counts in the value ratio column seem to indicate that the lack of reporting seems to be more of a Brazil issue. 

Or it could be that Paraguay is grossly misclassifying their exports. 

Either way, nearly half of Brazil-directed trade with Paraguay is poorly documented by one party or the other. This creates ripe ground for trade fraud. 

### Comparing data that can be compared

In order to ensure that we are only examining data that shows serious trade reporting discrepancies, it's necessary to filter out any rows where reported export quantities, values, or weights were within 15% of each other. 

By filtering out any rows that seems to demonstrate fairly consistent reporting, we are left with only rows where discrepancies are large enough to merit more thorough examination.

In [9]:
# Because we know that one of the rows in the full_rep subset has a string value in the wgt ratio column...
# ... we need to replace that with a zero for convenience sake. 
df_full_rep['wgt_ratio'].replace('kWh',0.0, inplace=True)

C:\Users\bfran\AppData\Local\Temp\ipykernel_5420\2564010108.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_full_rep['wgt_ratio'].replace('kWh',0.0, inplace=True)


In [10]:
# Convert both columns to numeric type, coercing any non-numeric values to NaN
df_full_rep['value_ratio'] = pd.to_numeric(df_full_rep['value_ratio'], errors='coerce')
df_full_rep['wgt_ratio'] = pd.to_numeric(df_full_rep['wgt_ratio'], errors='coerce')
df_full_rep['density_ratio'] = pd.to_numeric(df_full_rep['density_ratio'], errors='coerce')

# Filters out all rows where values are close to what we expect them to be. 
condition = ~((df_full_rep['value_ratio'] > 0.85) & (df_full_rep['value_ratio'] < 1.15) & 
              (df_full_rep['wgt_ratio'] > 0.85) & (df_full_rep['wgt_ratio'] < 1.15) &
              (df_full_rep['density_ratio'] > 0.85) & (df_full_rep['density_ratio'] < 1.15))

# Applies filter condition
df_full_rep = df_full_rep[condition]

C:\Users\bfran\AppData\Local\Temp\ipykernel_5420\3055919471.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_full_rep['value_ratio'] = pd.to_numeric(df_full_rep['value_ratio'], errors='coerce')
C:\Users\bfran\AppData\Local\Temp\ipykernel_5420\3055919471.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_full_rep['wgt_ratio'] = pd.to_numeric(df_full_rep['wgt_ratio'], errors='coerce')
C:\Users\bfran\AppData\Local\Temp\ipykernel_5420\3055919471.py:4: SettingWithCopyWarning: 
A value is trying to be s

In [11]:
df_full_rep.shape

(127, 18)

After filtering out all rows where the calculated ratios are what we expect them to be, we are left with 127 rows where ratio discrepancies seem to indicate possible trade fraud.

In [12]:
# Filters out all rows where values are close to what we expect them to be. 
condition_sm = ((df_full_rep['value_ratio'] < 0.85) & 
               (df_full_rep['wgt_ratio'] < 0.85) &
               ((df_full_rep['density_ratio'] < 1.15)))

In [13]:
sm = df_full_rep[condition_sm]

sm

hs_code     pry_x_q pry_x_unit     brz_m_q brz_m_unit  pry_x_net_wgt  \
5      20649   100016.00         kg    75014.00         kg      100016.00   
15     90300  1342970.00         kg  1099929.00         kg     1342970.00   
32    120890     9451.20         kg     6654.00         kg        9451.20   
53    200929    66300.00         kg    44200.00         kg       66300.00   
76    251200  1906500.00         kg  1379500.00         kg     1906500.00   
105   310590    66937.90         kg    51650.00         kg       66937.90   
162   391723        2.64         kg        1.00         kg           2.64   
168   391890   134189.00         kg    79811.00         kg      134189.00   
170   391990     1468.77         kg      157.65         kg        1468.77   
174   392043    49081.90         kg    31441.00         kg       49081.90   
223   420212        0.00         NR     3498.00          u       18921.43   
230   420292      154.31         kg       19.00         kg         154.31   
251   481910     3637.17         kg     2431.00         kg        3637.17   
252   481920     3396.83         kg     2390.00         kg        3396.83   
266   520922     7669.30         kg     4185.00         kg        7669.30   
311   590900      114.70         kg       91.37         kg         114.70   
343   610821   191957.18          u   288000.00          u       15732.60   
349   610910     1045.83          u      658.00          u         240.96   
362   620429        0.00         NR      624.00          u        1060.20   
418   680530    73224.00         kg    59136.00         kg       73224.00   
459   731829      192.40         kg       19.14         kg         192.40   
484   820559      167.00         kg       28.00         kg         167.00   
492   821000      344.94         kg       89.00         kg         344.94   
505   830210      152.20         kg        6.00         kg         152.20   
516   840991      134.36         kg      109.00         kg         134.36   
544   842199       73.56         kg       22.00         kg          73.56   
599   848210        0.00         NR       32.00          u         495.07   
600   848220        0.00         NR       50.00          u          35.06   
601   848240        0.00         NR        6.00          u           1.92   
610   848390     2604.47         kg     1408.00         kg        2604.47   
658   853610      269.34         kg       96.00         kg         269.34   
663   853690    21471.57         kg     4018.00         kg       21471.57   
730   920991      591.50         kg      400.00         kg         591.50   

     brz_m_net_wgt  pry_x_value  itic_rate  applied_tariff  adj_x_value  \
5         75014.00     92000.16       1.03            8.00     95082.17   
15      1099929.00   2018585.00       1.01            8.00   2046845.19   
32         6654.00     32647.20       1.04            8.00     33887.79   
53        44200.00    195630.03       1.02           11.20    200292.55   
76      1379500.00    322696.54       1.04            3.20    334355.47   
105       51650.00    250790.88       1.01            1.07    253006.20   
162           1.00       322.40       1.04           16.00       333.95   
168       79811.00    593512.06       1.01           12.80    598457.99   
170         157.65      8471.29       1.01           16.00      8554.59   
174       31441.00     72600.12       1.01           12.80     73531.82   
223       11301.00    153754.88       1.01           16.00    155215.55   
230          19.00     10643.78       1.01           35.00     10744.90   
251        2431.00      3559.81       1.06           12.80      3757.38   
252        2390.00     10828.20       1.06           12.80     11429.17   
266        4185.00     54445.04       1.01           26.00     55116.53   
311          91.37      3895.64       1.06           12.80      4146.91   
343        7652.00    384192.00       1.04           35.00    397766.78   
349         108.00      6821.45     

So 33 rows indicate signs of possible smuggling into Brazil. Other explanations of this phenomenon could be that the imported item was simply misclassified (meaning it should have called, say, HS 890111, but was instead called HS 891001), or it could be that the value of the product was understated by one party or another. If the product was undervalued by Paraguay, it could indicate that they are cutting some sort of deal to give someone extra product for free. Alternatively, if the product was undervalued by Brazil, it could be that someone in Brazil is trying to skim off some of the product or profits for themself. 

In [14]:
tariffs = pd.read_csv("brz_app_tariffs.csv")

tariffs.columns

Index(['reporter', 'year', 'hs_code', 'applied_tariff', 'commodity'], dtype='object')

In [15]:
columns = ['reporter','year','applied_tariff']

tariffs = tariffs.drop(columns=columns)

tariffs.shape

(16584, 2)

In [16]:
sm_codes = sm['hs_code']
sm_codes.shape

(33,)

In [22]:
sm_com = pd.merge(sm_codes, tariffs, on='hs_code', how='left')

In [26]:
sm_com.drop_duplicates(inplace=True)

In [27]:
sm_com

hs_code                                          commodity
0     20649       Edible offal of swine, frozen (excl. livers)
3     90300                                               Mate
6    120890  Flours and meal of oil seeds or oleaginous fru...
9    200929  Grapefruit juice, unfermented, Brix value > 20...
12   251200  Siliceous fossil meals, e.g. kieselguhr, tripo...
15   310590  Mineral or chemical fertilisers containing the...
18   391723  Rigid tubes, pipes and hoses, of polymers of v...
21   391890  Floor coverings of plastics, whether or not se...
24   391990  Self-adhesive plates, sheets, film, foil, tape...
27   392043  Plates, sheets, film, foil and strip, of non-c...
30   420212  Trunks, suitcases, vanity cases, executive-cas...
33   420292  Travelling-bags, insulated food or beverage ba...
36   481910  Cartons, boxes and cases, of corrugated paper ...
39   481920  Folding cartons, boxes and cases, of non-corru...
42   520922  Woven fabrics of cotton, containing >= 85% cot...
45   590900  Textile hosepiping and similar textile tubing,...
48   610821  Women's or girls' briefs and panties of cotton...
51   610910  T-shirts, singlets and other vests of cotton, ...
54   620429  Women's or girls' ensembles of textile materia...
57   680530  Natural or artificial abrasive powder or grain...
60   731829            Non-threaded articles, of iron or steel
63   820559  Hand tools, incl. glaziers' diamonds, of base ...
66   821000  Hand-operated mechanical devices, of base meta...
69   830210                 Hinges of all kinds, of base metal
72   840991  Parts suitable for use solely or principally w...
75   842199  Parts of machinery and apparatus for filtering...
78   848210                                      Ball bearings
81   848220  Tapered roller bearings, incl. cone and tapere...
84   848240  Needle roller bearings, incl. cage and needle ...
86   848240                             Needle roller bearings
87   848390  Toothed wheels, chain sprockets and other tran...
90   853610                     Fuses for a voltage <= 1.000 V
93   853690  Electrical apparatus for switching electrical ...
96   920991           Parts and accessories for pianos, n.e.s.